### PR TITLE
Implement alternative item-despawn-rate

### PR DESCRIPTION
**Pull request by kickash32**
This implements the feature request of #1710. If enabled, the specified Materials which are dropped as items will despawn at the chose alternate rate. Materials which are not selected will be unaffected regardless.

See #1710 for config setup (in paper.yml)

### CODE REVIEW

File: Spigot-Server-Patches/0399-Implement-alternative-item-despawn-rate.patch [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+@@ -109,6 +115,29 @@ public class PaperWorldConfig {
+         hardDespawnDistance = hardDespawnDistance*hardDespawnDistance;
+     }
+ 
++    public boolean altItemDespawnRateEnabled;
++    public Set<Material> altItemDespawnRateItems;
++    public int altItemDespawnRate;
++    private void altItemDespawnRate(){
++        altItemDespawnRateEnabled = getBoolean("alt-item-despawn-rate.enabled", false);
++        System.out.println(altItemDespawnRateEnabled);
```
Comment (MiniDigger) [inline Spigot-Server-Patches/0399-Implement-alternative-item-despawn-rate.patch]: remove this

File: Spigot-Server-Patches/0399-Implement-alternative-item-despawn-rate.patch [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
++        altItemDespawnRateItems = new HashSet<>();
++        for(Object input : getList("alt-item-despawn-rate.items", Arrays.asList("COBBLESTONE","STONE","DIRT","SANDSNOWBALL","NETHERRACK"))){
```
Comment (MiniDigger) [inline Spigot-Server-Patches/0399-Implement-alternative-item-despawn-rate.patch]: ~not sure if we want to supply default values like that, might be confusing to ppl when they update~
nvm, its disabled by default.

File: Spigot-Server-Patches/0399-Implement-alternative-item-despawn-rate.patch [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
++            try{
++                String name = (String)input;
++                altItemDespawnRateItems.add(Material.getMaterial(name));
++            }
++            catch (Exception e){
++                logError("Could not add item to altItemDespawnRateItems: "+e.getMessage());
```
Comment (MiniDigger) [inline Spigot-Server-Patches/0399-Implement-alternative-item-despawn-rate.patch]: how about?
```suggestion
+                logError("Could not add item '" + input + "'to altItemDespawnRateItems: " + e.getMessage());
```

File: Spigot-Server-Patches/0399-Implement-alternative-item-despawn-rate.patch (hunk 2 above)
Comment (MiniDigger) [reply Spigot-Server-Patches/0399-Implement-alternative-item-despawn-rate.patch]: I don't like the Object var here, getList is generic, you should be able to make java use the string type here.

**Comment by MiniDigger** [comment]
thinking more about this, wouldn't it make more sense to have a map with individual despawn rates per material? or would that be overkill?

**Comment by kickash32** [comment]
I see no downsides actually

File: Spigot-Server-Patches/0399-Implement-alternative-item-despawn-rate.patch (hunk 1 above)
Comment (Spottedleaf) [reply Spigot-Server-Patches/0399-Implement-alternative-item-despawn-rate.patch]: generally we use the log function in place of this, see other places where it's used in the config

File: Spigot-Server-Patches/0399-Implement-alternative-item-despawn-rate.patch [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+         hardDespawnDistance = hardDespawnDistance*hardDespawnDistance;
+     }
+ 
++    public boolean altItemDespawnRateEnabled;
++    public Set<Material> altItemDespawnRateItems;
++    public int altItemDespawnRate;
++    private void altItemDespawnRate(){
++        altItemDespawnRateEnabled = getBoolean("alt-item-despawn-rate.enabled", false);
++        System.out.println(altItemDespawnRateEnabled);
++        altItemDespawnRateItems = new HashSet<>();
```
Comment (Spottedleaf) [inline Spigot-Server-Patches/0399-Implement-alternative-item-despawn-rate.patch]: EnumSet

File: Spigot-Server-Patches/0399-Implement-alternative-item-despawn-rate.patch (hunk 1 above)
Comment (kickash32) [reply Spigot-Server-Patches/0399-Implement-alternative-item-despawn-rate.patch]: Yeah I knew that, I just forgot to take out this piece of debug code...
Its actually logged below

**Comment by kickash32** [comment]
Latest commit has added the requested changes.
But would it be worthwhile to use a map instead?

**Comment by kickash32** [comment]
Added per item config.
Config is now
```    
alt-item-despawn-rate:
      enabled: true
      items:
        COBBLESTONE: 300

File: Spigot-Server-Patches/0399-Implement-alternative-item-despawn-rate.patch [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
++import org.bukkit.configuration.ConfigurationSection;
+ import org.bukkit.configuration.file.YamlConfiguration;
+ import org.spigotmc.SpigotWorldConfig;
+ 
+@@ -109,6 +115,44 @@ public class PaperWorldConfig {
+         hardDespawnDistance = hardDespawnDistance*hardDespawnDistance;
+     }
+ 
++    public boolean altItemDespawnRateEnabled;
++    public Map<Material, Integer> altItemDespawnRateMap;
```
Comment (kashike) [inline Spigot-Server-Patches/0399-Implement-alternative-item-despawn-rate.patch]: You could use a fastutil map here, no?

Comment (Spottedleaf) [reply Spigot-Server-Patches/0399-Implement-alternative-item-despawn-rate.patch]: enummap is used

**Comment by kickash32** [comment]
updated for 1.14.4

**Comment by CeIebrimbor** [comment]
Reason this isn't getting pulled? Seems like an easy, straightforward PR.

**Comment by Spottedleaf** [comment]
time

File: Spigot-Server-Patches/0406-Implement-alternative-item-despawn-rate.patch [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+ // CraftBukkit end
+@@ -127,6 +128,19 @@ public class EntityItem extends Entity {
+                 }
+             }
+ 
++            // Paper start - alternative item despawn rate
++            Material material = this.getItemStack().getBukkitStack().getType();
++            if (world.paperConfig.altItemDespawnRateEnabled && world.paperConfig.altItemDespawnRateMap.containsKey(material)) {
++                if(!this.world.isClientSide && this.age >= world.paperConfig.altItemDespawnRateMap.get(material)) {
++                    if (org.bukkit.craftbukkit.event.CraftEventFactory.callItemDespawnEvent(this).isCancelled()) {
```
Comment (Spottedleaf) [inline Spigot-Server-Patches/0406-Implement-alternative-item-despawn-rate.patch]: avoid the duplication of death logic

(i.e use the spigot value if we don't have a config'd one)

File: Spigot-Server-Patches/0406-Implement-alternative-item-despawn-rate.patch [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+     }
++
++    public boolean altItemDespawnRateEnabled;
++    public Map<Material, Integer> altItemDespawnRateMap;
++    private void altItemDespawnRate() {
++        String path = "alt-item-despawn-rate";
++
++            altItemDespawnRateEnabled = getBoolean(path + ".enabled", false);
++
++            Map<Material, Integer> altItemDespawnRateMapDefault = new EnumMap<>(Material.class);
```
Comment (Spottedleaf) [inline Spigot-Server-Patches/0406-Implement-alternative-item-despawn-rate.patch]: the indentation here is wonky

**Comment by kickash32** [comment]
Fixed both issues.

File: Spigot-Server-Patches/0406-Implement-alternative-item-despawn-rate.patch [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+ import org.bukkit.event.entity.EntityPickupItemEvent;
+ import org.bukkit.event.player.PlayerPickupItemEvent;
+ // CraftBukkit end
+@@ -127,7 +128,8 @@ public class EntityItem extends Entity {
+                 }
+             }
+ 
+-            if (!this.world.isClientSide && this.age >= world.spigotConfig.itemDespawnRate) { // Spigot
++            Material material = this.getItemStack().getBukkitStack().getType(); // Paper
++            if (!this.world.isClientSide && this.age >= world.paperConfig.altItemDespawnRateMap.getOrDefault(material, world.spigotConfig.itemDespawnRate)) { // Spigot // Paper
```
Comment (Spottedleaf) [inline Spigot-Server-Patches/0406-Implement-alternative-item-despawn-rate.patch]: guard the get call by the enabled check

Comment (Spottedleaf) [reply Spigot-Server-Patches/0406-Implement-alternative-item-despawn-rate.patch]: nevermind

**Review by electronicboy — APPROVED** [review]

**Pull request merged**